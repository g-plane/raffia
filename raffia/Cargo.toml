[package]
name = "raffia"
version = "0.1.0"
edition = "2021"
authors = ["Pig Fang <g-plane@hotmail.com>"]
description = "One parser for CSS/SCSS/Sass/Less."
repository = "https://github.com/g-plane/raffia"
license = "MIT"

[dependencies]
raffia_macro = { path = "../raffia_macro", version = "0.1" }
serde = { version = "1.0", features = ["derive"], optional = true }
smallvec = { version = "1.9", features = ["const_generics", "serde"] }

[dev-dependencies]
codespan-reporting = "0.11"
criterion = "0.3"
insta = { version = "1.19", features = ["glob", "ron"] }
lightningcss = "1.0.0-alpha.35"
serde_json = "1.0"
swc_common = "0.29.0"
swc_css_ast = "0.114.0"
swc_css_parser = "0.123.0"

[features]
serialize = ["serde"]

[[bench]]
name = "parse_selector"
harness = false

[[bench]]
name = "parse_stylesheet"
harness = false
