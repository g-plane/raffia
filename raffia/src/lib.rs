//! Raffia is a parser can parse CSS, SCSS, Sass (indented syntax) and Less.
//!
//! ## Basic Usage
//!
//! This crate provides a simple API to get started.
//!
//! First, create a parser, give it the source code and specify the syntax,
//! then call the [`parse`](Parser::parse) method:
//!
//! ```rust
//! use raffia::{ast::Stylesheet, Parser, Syntax};
//!
//! let mut parser = Parser::new("a {}", Syntax::Css); // syntax can also be `Scss`, `Sass` or `Less`
//! let result = parser.parse::<Stylesheet>();
//! match result {
//!     Ok(ast) => {
//!         // parsed successfully
//!         println!("{:#?}", ast);
//!     }
//!     Err(error) => {
//!         // it failed, error message and position can be accessed via `error`
//!         println!("{:#?}", error);
//!     }
//! }
//! ```
//!
//! ## Advanced Usage
//!
//! ### More Parser Options
//!
//! If you need to control more parser options, you can use [`ParserBuilder`].
//!
//! For example, to collect comments:
//!
//! ```rust
//! use raffia::ParserBuilder;
//!
//! let mut comments = vec![];
//! let builder = ParserBuilder::new("/* comment */ a {}").comments(&mut comments);
//! let mut parser = builder.build();
//! ```
//!
//! By default, syntax is CSS when using parser builder. You can customize it:
//!
//! ```rust
//! use raffia::{ParserBuilder, Syntax};
//!
//! let builder = ParserBuilder::new("a {}").syntax(Syntax::Scss);
//! ```
//!
//! ### Parse Partial Structure
//!
//! Sometimes you don't want to parse a full stylesheet.
//! Say you only need to parse a qualified rule or even a single declaration.
//! All you need to do is to update the generics of the [`parse`](Parser::parse) method.
//!
//! ```rust
//! use raffia::{ast::QualifiedRule, Parser, Syntax};
//!
//! let mut parser = Parser::new("a {}", Syntax::Css);
//! parser.parse::<QualifiedRule>();
//! ```
//!
//! and
//!
//! ```rust
//! use raffia::{ast::Declaration, Parser, Syntax};
//!
//! let mut parser = Parser::new("color: green", Syntax::Css);
//! parser.parse::<Declaration>();
//! ```
//!
//! Not all AST nodes support the usage above;
//! technically, those nodes that implement [`Parse`] trait are supported.
//!
//! ### Retrieve Recoverable Errors
//!
//! There may be some recoverable errors which doesn't affect on producing AST.
//! To retrieve those errors, use [`recoverable_errors`](Parser::recoverable_errors).
//!
//! ```rust
//! use raffia::{ast::Stylesheet, Parser, Syntax};
//!
//! let mut parser = Parser::new("@keyframes kf { invalid {} }", Syntax::Css);
//! let result = parser.parse::<Stylesheet>();
//! assert!(result.is_ok());
//! println!("{:?}", parser.recoverable_errors());
//! ```
//!
//! ## Serialization
//!
//! Produced AST can be serialized by Serde, but this feature is disabled by default.
//! You need to enable feature `serialize` manually:
//!
//! ```toml
//! raffia = { version = "*", features = ["serialize"] }
//! ```
//!
//! Then you can pass AST to Serde.
//!
//! Note that Raffia only supports serialization. Deserialization isn't supported.

pub use config::{ParserOptions, Syntax};
pub use parser::{Parse, Parser, ParserBuilder};
pub use pos::{Span, Spanned};
pub use span_ignored_eq::SpanIgnoredEq;
pub use tokenizer::token;

pub mod ast;
mod config;
pub mod error;
mod parser;
pub mod pos;
mod span_ignored_eq;
mod tokenizer;
mod util;
