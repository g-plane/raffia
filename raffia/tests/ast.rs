use codespan_reporting::{
    diagnostic::{Diagnostic, Label},
    files::SimpleFile,
    term::{self, termcolor::Buffer},
};
use insta::{assert_ron_snapshot, glob, Settings};
use raffia::{ast::Stylesheet, Parser, Syntax};
use std::fs;

#[test]
fn ast_snapshot() {
    glob!("ast/**/*.{css,scss,sass,less}", |path| {
        let code = fs::read_to_string(path).unwrap();
        let syntax = match path.extension().unwrap().to_str().unwrap() {
            "css" => Syntax::Css,
            "scss" => Syntax::Scss,
            "sass" => Syntax::Sass,
            "less" => Syntax::Less,
            _ => unreachable!("unknown file extension"),
        };
        let mut parser = Parser::new(&code, syntax);
        let ast = match parser.parse::<Stylesheet>() {
            Ok(ast) => {
                let recoverable_errors = parser.recoverable_errors();
                assert!(
                    recoverable_errors.is_empty(),
                    "'{}' has recoverable errors: {recoverable_errors:?}",
                    path.file_name().unwrap().to_str().unwrap(),
                );
                ast
            }
            Err(error) => {
                let file = SimpleFile::new(path.file_name().unwrap().to_str().unwrap(), &code);
                let diagnostic = Diagnostic::error()
                    .with_message(error.kind.to_string())
                    .with_labels(vec![Label::primary((), error.span.start..error.span.end)]);
                let mut buffer = Buffer::ansi();
                let config = term::Config::default();
                term::emit(&mut buffer, &config, &file, &diagnostic).unwrap();
                panic!("\n{}", String::from_utf8(buffer.into_inner()).unwrap());
            }
        };

        let mut settings = Settings::clone_current();
        settings.set_snapshot_path(path.parent().unwrap());
        settings.remove_snapshot_suffix();
        settings.set_prepend_module_to_snapshot(false);
        settings.remove_input_file();
        settings.set_omit_expression(true);
        settings.remove_input_file();
        settings.remove_info();
        settings.bind(|| {
            let name = path.file_stem().unwrap().to_str().unwrap();
            assert_ron_snapshot!(name, ast);
        });
    });
}
