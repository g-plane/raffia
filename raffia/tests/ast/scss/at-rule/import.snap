---
source: raffia/tests/ast.rs
---
Stylesheet(
  type: "Stylesheet",
  statements: [
    AtRule(
      type: "AtRule",
      name: Ident(
        type: "Ident",
        name: "import",
        raw: "import",
        span: Span(
          start: 1,
          end: 7,
        ),
      ),
      prelude: Some(ImportPrelude(
        type: "ImportPrelude",
        href: Str(
          type: "Str",
          value: "mod",
          raw: "\"mod\"",
          span: Span(
            start: 8,
            end: 13,
          ),
        ),
        layer: None,
        supports: None,
        media: None,
        span: Span(
          start: 8,
          end: 13,
        ),
      )),
      block: None,
      span: Span(
        start: 0,
        end: 13,
      ),
    ),
    AtRule(
      type: "AtRule",
      name: Ident(
        type: "Ident",
        name: "import",
        raw: "import",
        span: Span(
          start: 16,
          end: 22,
        ),
      ),
      prelude: Some(ImportPrelude(
        type: "ImportPrelude",
        href: Str(
          type: "Str",
          value: "landscape",
          raw: "\"landscape\"",
          span: Span(
            start: 23,
            end: 34,
          ),
        ),
        layer: None,
        supports: None,
        media: Some(MediaQueryList(
          type: "MediaQueryList",
          queries: [
            MediaQueryWithType(
              type: "MediaQueryWithType",
              modifier: None,
              mediaType: Ident(
                type: "Ident",
                name: "screen",
                raw: "screen",
                span: Span(
                  start: 35,
                  end: 41,
                ),
              ),
              condition: Some(MediaConditionAfterMediaType(
                type: "MediaConditionAfterMediaType",
                and: Ident(
                  type: "Ident",
                  name: "and",
                  raw: "and",
                  span: Span(
                    start: 42,
                    end: 45,
                  ),
                ),
                condition: MediaCondition(
                  type: "MediaCondition",
                  conditions: [
                    MediaInParens(
                      type: "MediaInParens",
                      kind: MediaFeaturePlain(
                        type: "MediaFeaturePlain",
                        name: Ident(
                          type: "Ident",
                          name: "orientation",
                          raw: "orientation",
                          span: Span(
                            start: 47,
                            end: 58,
                          ),
                        ),
                        colonSpan: Span(
                          start: 58,
                          end: 59,
                        ),
                        value: Ident(
                          type: "Ident",
                          name: "landscape",
                          raw: "landscape",
                          span: Span(
                            start: 60,
                            end: 69,
                          ),
                        ),
                        span: Span(
                          start: 47,
                          end: 69,
                        ),
                      ),
                      span: Span(
                        start: 46,
                        end: 70,
                      ),
                    ),
                  ],
                  span: Span(
                    start: 46,
                    end: 70,
                  ),
                ),
                span: Span(
                  start: 42,
                  end: 70,
                ),
              )),
              span: Span(
                start: 35,
                end: 70,
              ),
            ),
          ],
          commaSpans: [],
          span: Span(
            start: 35,
            end: 70,
          ),
        )),
        span: Span(
          start: 23,
          end: 70,
        ),
      )),
      block: None,
      span: Span(
        start: 15,
        end: 70,
      ),
    ),
  ],
  span: Span(
    start: 0,
    end: 72,
  ),
)
